src/common.c

```diff
@@ -210,15 +210,15 @@ sr_ds_handle_init(struct sr_ds_handle_s **ds_handles, uint32_t *ds_handle_count)
     if (!plugins_dir[0]) {
         /* get plugins dir from environment variable, or use default one */
         tmp = getenv("SR_PLUGINS_PATH");
-        if (tmp) {
-            if ((len = strlen(tmp)) < SR_PATH_MAX) {
-                strncpy(plugins_dir, tmp, len);
-            } else {
-                sr_errinfo_new(&err_info, SR_ERR_INVAL_ARG, "SR_PLUGINS_PATH (%s) cannot be longer than %u.", tmp, SR_PATH_MAX);
-                goto cleanup;
-            }
+        if (!tmp) {
+            tmp = SR_PLG_PATH;
+        }
+
+        if ((len = strlen(tmp)) < SR_PATH_MAX) {
+            strncpy(plugins_dir, tmp, SR_PATH_MAX);
```

Comment:
Why do you insist on using `SR_PATH_MAX` for the length when such a length is invalid? I insist on using a valid length instead, otherwise this line is confusing for no reason at all. Please use either `len` or `SR_PATH_MAX - 1` is fine as well, where `len` is not used.

Reply:
I was trying to address a warning you can see here [build-log-link](https://productionresultssa5.blob.core.windows.net/actions-results/79812071-a81b-4be6-9587-95fb984355c3/workflow-job-run-dfb3d9d1-7d78-56db-b388-e25a8a802d50/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-10-15T08%3A27%3A45Z&sig=%2FOyBzxeGFCRZfMIO7buTXuOnAnAKI5%2BKGD4rFch42IM%3D&ske=2024-10-15T17%3A37%3A54Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-10-15T05%3A37%3A54Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-08-04&sp=r&spr=https&sr=b&st=2024-10-15T08%3A17%3A40Z&sv=2024-08-04).

```C
2024-10-11T13:40:16.5830026Z In function ‘strncpy’,
2024-10-11T13:40:16.5831926Z     inlined from ‘sr_ds_handle_init’ at /home/runner/work/sysrepo/sysrepo/pkg/build/pkgs/ubuntu-22.04/sysrepo_2.11.10-1/sysrepo-2.11.10/src/common.c:215:17:
2024-10-11T13:40:16.5835253Z /usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10: warning: ‘__strncpy_chk’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
2024-10-11T13:40:16.5837328Z    95 |   return __builtin___strncpy_chk (__dest, __src, __len,
2024-10-11T13:40:16.5838277Z       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-10-11T13:40:16.5839258Z    96 |                                   __glibc_objsize (__dest));
2024-10-11T13:40:16.5840023Z       |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~
2024-10-11T13:40:16.5841948Z /home/runner/work/sysrepo/sysrepo/pkg/build/pkgs/ubuntu-22.04/sysrepo_2.11.10-1/sysrepo-2.11.10/src/common.c: In function ‘sr_ds_handle_init’:
2024-10-11T13:40:16.5844951Z /home/runner/work/sysrepo/sysrepo/pkg/build/pkgs/ubuntu-22.04/sysrepo_2.11.10-1/sysrepo-2.11.10/src/common.c:214:24: note: length computed here
2024-10-11T13:40:16.5846622Z   214 |             if ((len = strlen(tmp)) < SR_PATH_MAX) {
```

Using `len` will not copy the null termination character from `tmp`.
When using `SR_PATH_MAX-1`, if `strlen(tmp) == SR_PATH_MAX-1`, then again `strncpy` will not copy the null termination character.

However in this case, since `plugins_dir` is `static`, the last byte is already 0 and does not need to be copied, So I guess it really is a false-positive.

If it is confusing to use `SR_PATH_MAX`, I could use `len+1`, would that be better?

Reply:
Fair enough, such a trivial task and so difficult to write it correctly. What about just using `strcpy()`, do you think the compiler/static analysis would mind?

Reply:
`strcpy` also causes problems for analyzers such as 
> This 'call to strcpy' with input from may overflow the destination. CodeQL
The program performs a buffer copy or write operation with no upper limit on the size of the copy. An unexpectedly long input that reaches this code will cause the buffer to overflow. In addition to causing program instability, techniques exist which may allow an attacker to use this vulnerability to execute arbitrary code.

Reply:
Yeah, but I would expect them to be able to figure out the length is checked before. Whatever, I would then prefer using `snprintf()` to `strncpy()` because you can use `SR_PATH_MAX` as the destination buffer length.